### ISSUE TITLE
Can I compile a single file (bundled) version of Tape to work with Rollup in the browser?

### ISSUE THREAD

**Issue by David-Else** [substack/tape#410]
Hi, I am have a nightmare trying to make Tape work with Rollup. Rollup seems to require all sorts of ancient abandoned plugins to try to pull Tape into the bundle (due to node 'builtins') to make it work in the browser... and I have given up. 

I am trying to do something like this guide demonstrates, but with Rollup:

https://medium.com/@arnaudrinquin/frictionless-unit-testing-in-javascript-with-browser-tap-6ac2cea89a59

I am a newb, and don't really understand Node. I know that Tape uses Node modules, and that browserify can recreate them for you when you run in the browser, so I tried this in a desperate attempt:

`/node_modules/.bin/browserify /node_modules/tape/bin/tape -o pleasework.js`

and got a file, but `import test from "../src/pleasework"; `just results in `TypeError: pleasework is not a function`

I am probably really annoying everyone with something so basic, but I am very frustrated (with my own ignorance!)

So to sum up, I want to use Rollup rather than Browserify. Rollup does not work, can I compile the file I need using Browserify on the command line and just import it with Rollup and do my tests?

Many many thanks for your time reading this!

**Comment by ljharb** [comment]
If rollup doesn't work, why do you want to use rollup? Rollup's primary claimed benefit is bundle size; but bundle size doesn't matter for testing.

Separately, you don't want to use the `tape` binary unless you're in node; just invoking the files directly should be sufficient.

**Comment by David-Else** [comment]
Thanks for getting back to me! Regarding using the files directly, this works great:

```
var test = require('tape'); // assign the tape library to the variable "test"

function sum(a, b) {
  console.log("hello");
  return a + b;
}

test('sum should return the addition of two numbers', function (t) {
  t.equal(3, sum(1, 2)); // make this test pass by completing the add function!
  t.end();
});

TYPE ON THE CONSOLE

../node_modules/tape/bin/tape test1.js 
```

But I don't want all my code in the `test1.js` file. As soon as I try to import the `sum` function into the file, it stops working and says that the function is not a function:

```
var test = require('tape'); // assign the tape library to the variable "test"
var sum = require('../test/function');

test('sum should return the addition of two numbers', function (t) {
  t.equal(3, sum(1, 2)); // make this test pass by completing the add function!
  t.end();
});

SEPARATE FILE function.js:

function sum(a, b) {
  console.log("hello");
  return a + b;
}
```
I keep running up against this problem, could you have mercy on me and tell me how to make it work?!

Ideally I want to just require('../dist/bundle.js') and do all my tests on that final file by pulling it into the test file (like I have tried above with `require('../test/function')` and failed.

Will node understand all my browser API code? I have literally spent all day on this, either I am OCD or JS development is insanely complex.

**Comment by ljharb** [comment]
Ah. You're not *exporting* `sum` from the first file. Try `module.exports = sum;` - that will make `require('../test/function')` provide `sum`.

**Comment by David-Else** [comment]
Cheers, that works!

Now my final problem is how do I do the same thing to `require` my final bundle.js IIFE file that Rollup creates?

Maybe I am missing some obvious workflow thing here? How do I import an IIFE and access all it's functions to test? There is no `module.exports = sum;` type command that could do this from outside of the file?

Basically I have my final bundled js file and want to run tests on it. I imagine my tests file would look something like:

```
var test = require('tape'); 
var entireJSFile = require('../dist/bundle.js');

test('test1');
test('test2');
etc...
```
Is this the best way to do it? How would you achieve my goal? THANKS.

**Comment by ljharb** [comment]
(I'm going to close this, since this isn't a tape question; rather it's a question about how module systems work in JS)

When you're using a module system, you basically never have an IIFE ever. You browserify (or webpack, or rollup) your entry point, and every file is written as a node-style CJS module - requiring what it needs, and `module.exports =`ing if it wants to export a single value.

Separately, I'm not sure why you'd need to run tests on your bundled file at all - you'd just require the actual piece you need. You don't need to test your bundler.

**Comment by David-Else** [comment]
Thanks for all your help, I am going to have to sit down and work this all out properly myself and break down the problem.
